exfat/fuse/mount.exfat-fuse.8
.\" Copyright (C) 2010-2016  Andrew Nayenko
.\"
.TH EXFAT-FUSE 8 "November 2015"
.SH NAME
mount.exfat-fuse \- mount an exFAT file system
.SH SYNOPSIS
.B mount.exfat-fuse
[
.B \-d
]
[
.B \-n
]
[
.B \-o
.I options
]
[
.B \-V
]
[
.B \-v
]
.I device dir

.SH DESCRIPTION
.B mount.exfat-fuse
is a free exFAT file system implementation with write support. exFAT is a
simple file system created by Microsoft. It is intended to replace FAT32
removing some of its limitations. exFAT is a standard FS for SDXC memory
cards.

.SH COMMAND LINE OPTIONS
Command line options available:
.TP
.BI \-d
Enable debug logging and do not detach from shell.
.TP
.BI \-n
Ignored.
.TP
.BI \-o " options"
File system specific options. For more details see
.B FILE SYSTEM OPTIONS
section below.
.TP
.BI \-V
Print version and copyright.
.TP
.BI \-v
Ignored.

.SH FILE SYSTEM OPTIONS
.TP
.BI umask= value
Set the umask (the bitmask of the permissions that are
.B not
present, in octal).
The default is 0.
.TP
.BI dmask= value
Set the umask for directories only.
.TP
.BI fmask= value
Set the umask for files only.
.TP
.BI uid= n
Set the owner for all files and directories.
The default is the owner of the current process.
.TP
.BI gid= n
Set the group for all files and directories.
The default is the group of the current process.
.TP
.BI ro
Mount the file system in read only mode.
.TP
.BI noatime
Do not update access time when file is read.

.SH EXIT CODES
Zero is returned on successful mount. Any other code means an error.

.SH BUGS
exFAT is a case-insensitive file system. Some things can behave unexpectedly,
e.g. directory renaming that changes only case of some characters:

.B \t$ mv FOO Foo
.br
.B \tmv: cannot move \(cqFOO\(cq to a subdirectory of itself, \(cqFoo/FOO\(cq

This happens because
.B mv
finds that destination exists (for case-insensitive file
systems
.B FOO
and
.B Foo
are the same thing) and adds source basename to the destination. The file
system gets
.B rename(\(dqFOO\(dq,\ \(dqFoo/FOO\(dq)
syscall and returns an error.

.SH AUTHOR
Andrew Nayenko

.SH SEE ALSO
.BR mount (8)


exfat/ChangeLog
1.3.0 (2018-09-15)

* exfatfsck can now repair some errors.
* Added experimental Android support for exfat-utils [liminghao, LongPingWEI,
Pablo Mendez Hernandez, Pierre-Hugues Husson].
* Cleaned up FUSE code preparing for FUSE 3 support.
* Removed OpenBSD support as it does not handle -o option in fuse_main().
* Re-introduced FreeBSD support [Oleksii Samorukov].
* Fixed DragonFly BSD support [Tomohiro Kusumi].
* dirent->d_type in now filled on readdir() [Mark Browning].

1.2.8 (2018-02-03)

* Fixed new files or directories creation in the root directory: ensure there
are no 0x00 entries before valid ones; otherwise Windows can corrupt them.
* Fixed compilation on GNU/HURD platform.

1.2.7 (2017-06-05)

* Fixed handling of two last clusters: operations with files that occupy these
clusters could fail.
* Fixed crash when started with stdin, stdout or stderr closed.

1.2.6 (2017-01-28)

* Operations with directories (except initial listing) now make less
read/write system calls.
* Fixed handling of files with optional tail entries (0xe0-0xff): videoclip
files created by Sony cameras were missing.
* Write operations now correctly return ENOSPC (instead of EIO) when there is
no free disk space left.
* Fixed max file name length: it's 255 16-bit code units (not 256).

1.2.5 (2016-12-05)

* Added an option for dumpexfat to show file fragments [Daniel Drake].
* Fixed crash when directory starts with an invalid cluster.
* Daylight saving time in now properly reflected in file timestamps.

1.2.4 (2016-06-03)

* Fixed wrong files names hashes when upper case table is compressed.
* Man pages are now installed by default.
* Commas and backslashes in device names are now escaped.

1.2.3 (2015-12-19)

* Fixed clusters loss when file renaming replaces target.

1.2.2 (2015-11-09)

* Improved reliability in case of a sudden unplug: FS will be in a clean state
after closing all files and performing sync(1).
* Fixed compilation on Debian GNU/kFreeBSD and GNU/Hurd platforms.
* Updated mount.exfat-fuse man page.

1.2.1 (2015-09-24)

* Fixed compatibility with Zalman VE-200: now newly created directories do not
have archive bit set.
* Fixed heap corruption: malformed FS can use invalid sector or cluster size.
* Fixed hang on mount: malformed FS can have cyclic references in the clusters
map.

1.2.0 (2015-08-26)

* Switched from SCons to autotools.
* Added musl libc support [Brendan Heading].
* Worked around "FS is larger than device" error for memory cards formatted by
Panasonic Lumix cameras.
* Worked around "unknown entry type 0xe1" error for memory cards formatted by
Sony cameras.

1.1.1 (2014-11-15)

* Fixed mkfs crash on some sectors-per-cluster (-s option) values.

1.1.0 (2014-07-08)

* Relicensed the project from GPLv3+ to GPLv2+.
* OpenBSD support [Helg Bredow].
* Improved I/O errors handling.
* Implemented fsync() and fsyncdir().
* Fixed crash on Mac OS X 10.5 caused by non-standard use of realpath(). Also
fixed TrueCrypt disks unmounting.
* Avoid extra erase on writes to the end of a file. This should improve linear
write speed.
* Allow arbitrary changing of lower 9 bits of mode. Allow owner/group changing
to the same owner/group. This fixes rsync.
* Fixed buffers overflows when handling lengthy file names.
* Fixed "real size does not equal to size" error on volumes with pagefile.sys.
* Fixed negative IUsed in "df -i" output.

1.0.1 (2013-02-02)

* Fixed unexpected removal of a directory if it is moved into itself.
* Fixed "Operation not permitted" error on reading an empty file.

1.0.0 (2013-01-19)

* Fixed crash when renaming a file within a single directory and a new name
differs only in case.
* Fixed clusters allocation: a cluster beyond valid clusters range could be
allocated.
* Fixed crash when a volume is unmounted while some files are open.
* SConscript now respects AR and RANLIB environment variables.
* Improved error handling.

Linux:

* Enabled big_writes. This improves write speed (larger block size means less
switches between kernel- and user-space).
* Do BLKROGET ioctl to make sure the device is not read-only: after
"blockdev --setro" kernel still allows to open the device in read-write mode
but fails writes.

OS X:

* Fixed OS X 10.8 support.
* Switched to 64-bit inode numbers (now Mac OS X 10.5 or later is required).
* Switched from unmaintained MacFUSE to OSXFUSE (http://osxfuse.github.com).
* Fixed device size detection. Now mkfs works.
* Workarounded some utilities failures due to missing chmod() support.
* Disabled (senseless) permission checks made by FUSE.

0.9.8 (2012-08-09)

* The mkfs utility can now create huge file systems (up to several exabytes).
* Fixed handling of characters beyond Basic Multilingual Plane.
* Echo messages to syslog only if stderr is not connected to a terminal.

0.9.7 (2012-03-08)

* Out-of-the-box FreeBSD support (via ublio library).
* Fixed "missing EOD entry" error (could happen while reading directory that
consists of several clusters).
* Fixed interpretation of minutes field in files timestamps (minutes could be
displayed incorrectly).
* Fixed mtime seconds field initialization for newly created file (mtime could
be 1 sec less than creation time).
* SConscript now respects CC, CCFLAGS and LDFLAGS environment variables.

0.9.6 (2012-01-14)

* Fixed write performance regression introduced in 0.9.4.
* Mount in read-only mode if the device is write-protected.
* Set ctime to mtime to ensure we don't break programs that rely on ctime
(e.g. rsync considered that all files are outdated) [Eldad Zack].
* Indicate that FS in not clean when it was not cleanly unmounted.
* Utilities are now compatible with GNU/Hurd.
* Fixed several memory leaks that could occur on error handling paths.
* Improved handling of corrupted file systems.

0.9.5 (2011-05-15)

* Fixed erasing of the root directory cluster when creating a new FS with
mkexfatfs. This bug could cause mkexfatfs to produce invalid FS.
* Utilities are not linked with libfuse anymore.

exfat/libexfat/mount.c
/*
	mount.c (22.10.09)
	exFAT file system implementation library.

	Free exFAT implementation.
	Copyright (C) 2010-2018  Andrew Nayenko

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License along
	with this program; if not, write to the Free Software Foundation, Inc.,
	51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "exfat.h"
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <inttypes.h>
#include <unistd.h>
#include <sys/types.h>

static uint64_t rootdir_size(const struct exfat* ef)
{
	uint32_t clusters = 0;
	uint32_t clusters_max = le32_to_cpu(ef->sb->cluster_count);
	cluster_t rootdir_cluster = le32_to_cpu(ef->sb->rootdir_cluster);

	/* Iterate all clusters of the root directory to calculate its size.
	   It can't be contiguous because there is no flag to indicate this. */
	do
	{
		if (clusters == clusters_max) /* infinite loop detected */
		{
			exfat_error("root directory cannot occupy all %d clusters",
					clusters);
			return 0;
		}
		if (CLUSTER_INVALID(*ef->sb, rootdir_cluster))
		{
			exfat_error("bad cluster %#x while reading root directory",
					rootdir_cluster);
			return 0;
		}
		rootdir_cluster = exfat_next_cluster(ef, ef->root, rootdir_cluster);
		clusters++;
	}
	while (rootdir_cluster != EXFAT_CLUSTER_END);

	return (uint64_t) clusters * CLUSTER_SIZE(*ef->sb);
}

static const char* get_option(const char* options, const char* option_name)
{
	const char* p;
	size_t length = strlen(option_name);

	for (p = strstr(options, option_name); p; p = strstr(p + 1, option_name))
		if ((p == options || p[-1] == ',') && p[length] == '=')
			return p + length + 1;
	return NULL;
}

static int get_int_option(const char* options, const char* option_name,
		int base, int default_value)
{
	const char* p = get_option(options, option_name);

	if (p == NULL)
		return default_value;
	return strtol(p, NULL, base);
}

static void parse_options(struct exfat* ef, const char* options)
{
	int opt_umask;

	opt_umask = get_int_option(options, "umask", 8, 0);
	ef->dmask = get_int_option(options, "dmask", 8, opt_umask);
	ef->fmask = get_int_option(options, "fmask", 8, opt_umask);

	ef->uid = get_int_option(options, "uid", 10, geteuid());
	ef->gid = get_int_option(options, "gid", 10, getegid());

	ef->noatime = exfat_match_option(options, "noatime");

	switch (get_int_option(options, "repair", 10, 0))
	{
	case 1:
		ef->repair = EXFAT_REPAIR_ASK;
		break;
	case 2:
		ef->repair = EXFAT_REPAIR_YES;
		break;
	default:
		ef->repair = EXFAT_REPAIR_NO;
		break;
	}
}

static bool verify_vbr_checksum(const struct exfat* ef, void* sector)
{
	off_t sector_size = SECTOR_SIZE(*ef->sb);
	uint32_t vbr_checksum;
	size_t i;

	if (exfat_pread(ef->dev, sector, sector_size, 0) < 0)
	{
		exfat_error("failed to read boot sector");
		return false;
	}
	vbr_checksum = exfat_vbr_start_checksum(sector, sector_size);
	for (i = 1; i < 11; i++)
	{
		if (exfat_pread(ef->dev, sector, sector_size, i * sector_size) < 0)
		{
			exfat_error("failed to read VBR sector");
			return false;
		}
		vbr_checksum = exfat_vbr_add_checksum(sector, sector_size,
				vbr_checksum);
	}
	if (exfat_pread(ef->dev, sector, sector_size, i * sector_size) < 0)
	{
		exfat_error("failed to read VBR checksum sector");
		return false;
	}
	for (i = 0; i < sector_size / sizeof(vbr_checksum); i++)
		if (le32_to_cpu(((const le32_t*) sector)[i]) != vbr_checksum)
		{
			exfat_error("invalid VBR checksum 0x%x (expected 0x%x)",
					le32_to_cpu(((const le32_t*) sector)[i]), vbr_checksum);
			if (!EXFAT_REPAIR(invalid_vbr_checksum, ef, sector, vbr_checksum))
				return false;
		}
	return true;
}

static int commit_super_block(const struct exfat* ef)
{
	if (exfat_pwrite(ef->dev, ef->sb, sizeof(struct exfat_super_block), 0) < 0)
	{
		exfat_error("failed to write super block");
		return 1;
	}
	return exfat_fsync(ef->dev);
}

int exfat_soil_super_block(const struct exfat* ef)
{
	if (ef->ro)
		return 0;

	ef->sb->volume_state = cpu_to_le16(
			le16_to_cpu(ef->sb->volume_state) | EXFAT_STATE_MOUNTED);
	return commit_super_block(ef);
}

static void exfat_free(struct exfat* ef)
{
	exfat_close(ef->dev);	/* first of all, close the descriptor */
	ef->dev = NULL;			/* struct exfat_dev is freed by exfat_close() */
	free(ef->root);
	ef->root = NULL;
	free(ef->zero_cluster);
	ef->zero_cluster = NULL;
	free(ef->cmap.chunk);
	ef->cmap.chunk = NULL;
	free(ef->upcase);
	ef->upcase = NULL;
	free(ef->sb);
	ef->sb = NULL;
}

int exfat_mount(struct exfat* ef, const char* spec, const char* options)
{
	int rc;
	enum exfat_mode mode;

	exfat_tzset();
	memset(ef, 0, sizeof(struct exfat));

	parse_options(ef, options);

	if (exfat_match_option(options, "ro"))
		mode = EXFAT_MODE_RO;
	else if (exfat_match_option(options, "ro_fallback"))
		mode = EXFAT_MODE_ANY;
	else
		mode = EXFAT_MODE_RW;
	ef->dev = exfat_open(spec, mode);
	if (ef->dev == NULL)
		return -EIO;
	if (exfat_get_mode(ef->dev) == EXFAT_MODE_RO)
	{
		if (mode == EXFAT_MODE_ANY)
			ef->ro = -1;
		else
			ef->ro = 1;
	}

	ef->sb = malloc(sizeof(struct exfat_super_block));
	if (ef->sb == NULL)
	{
		exfat_error("failed to allocate memory for the super block");
		exfat_free(ef);
		return -ENOMEM;
	}
	memset(ef->sb, 0, sizeof(struct exfat_super_block));

	if (exfat_pread(ef->dev, ef->sb, sizeof(struct exfat_super_block), 0) < 0)
	{
		exfat_error("failed to read boot sector");
		exfat_free(ef);
		return -EIO;
	}
	if (memcmp(ef->sb->oem_name, "EXFAT   ", 8) != 0)
	{
		exfat_error("exFAT file system is not found");
		exfat_free(ef);
		return -EIO;
	}
	/* sector cannot be smaller than 512 bytes */
	if (ef->sb->sector_bits < 9)
	{
		exfat_error("too small sector size: 2^%hhd", ef->sb->sector_bits);
		exfat_free(ef);
		return -EIO;
	}
	/* officially exFAT supports cluster size up to 32 MB */
	if ((int) ef->sb->sector_bits + (int) ef->sb->spc_bits > 25)
	{
		exfat_error("too big cluster size: 2^(%hhd+%hhd)",
				ef->sb->sector_bits, ef->sb->spc_bits);
		exfat_free(ef);
		return -EIO;
	}
	ef->zero_cluster = malloc(CLUSTER_SIZE(*ef->sb));
	if (ef->zero_cluster == NULL)
	{
		exfat_error("failed to allocate zero sector");
		exfat_free(ef);
		return -ENOMEM;
	}
	/* use zero_cluster as a temporary buffer for VBR checksum verification */
	if (!verify_vbr_checksum(ef, ef->zero_cluster))
	{
		exfat_free(ef);
		return -EIO;
	}
	memset(ef->zero_cluster, 0, CLUSTER_SIZE(*ef->sb));
	if (ef->sb->version.major != 1 || ef->sb->version.minor != 0)
	{
		exfat_error("unsupported exFAT version: %hhu.%hhu",
				ef->sb->version.major, ef->sb->version.minor);
		exfat_free(ef);
		return -EIO;
	}
	if (ef->sb->fat_count != 1)
	{
		exfat_error("unsupported FAT count: %hhu", ef->sb->fat_count);
		exfat_free(ef);
		return -EIO;
	}
	if (le64_to_cpu(ef->sb->sector_count) * SECTOR_SIZE(*ef->sb) >
			(uint64_t) exfat_get_size(ef->dev))
	{
		/* this can cause I/O errors later but we don't fail mounting to let
		   user rescue data */
		exfat_warn("file system in sectors is larger than device: "
				"%"PRIu64" * %d > %"PRIu64,
				le64_to_cpu(ef->sb->sector_count), SECTOR_SIZE(*ef->sb),
				exfat_get_size(ef->dev));
	}
	if ((off_t) le32_to_cpu(ef->sb->cluster_count) * CLUSTER_SIZE(*ef->sb) >
			exfat_get_size(ef->dev))
	{
		exfat_error("file system in clusters is larger than device: "
				"%u * %d > %"PRIu64,
				le32_to_cpu(ef->sb->cluster_count), CLUSTER_SIZE(*ef->sb),
				exfat_get_size(ef->dev));
		exfat_free(ef);
		return -EIO;
	}
	if (le16_to_cpu(ef->sb->volume_state) & EXFAT_STATE_MOUNTED)
		exfat_warn("volume was not unmounted cleanly");

	ef->root = malloc(sizeof(struct exfat_node));
	if (ef->root == NULL)
	{
		exfat_error("failed to allocate root node");
		exfat_free(ef);
		return -ENOMEM;
	}
	memset(ef->root, 0, sizeof(struct exfat_node));
	ef->root->attrib = EXFAT_ATTRIB_DIR;
	ef->root->start_cluster = le32_to_cpu(ef->sb->rootdir_cluster);
	ef->root->fptr_cluster = ef->root->start_cluster;
	ef->root->name[0] = cpu_to_le16('\0');
	ef->root->size = rootdir_size(ef);
	if (ef->root->size == 0)
	{
		exfat_free(ef);
		return -EIO;
	}
	/* exFAT does not have time attributes for the root directory */
	ef->root->mtime = 0;
	ef->root->atime = 0;
	/* always keep at least 1 reference to the root node */
	exfat_get_node(ef->root);

	rc = exfat_cache_directory(ef, ef->root);
	if (rc != 0)
		goto error;
	if (ef->upcase == NULL)
	{
		exfat_error("upcase table is not found");
		goto error;
	}
	if (ef->cmap.chunk == NULL)
	{
		exfat_error("clusters bitmap is not found");
		goto error;
	}

	return 0;

error:
	exfat_put_node(ef, ef->root);
	exfat_reset_cache(ef);
	exfat_free(ef);
	return -EIO;
}

static void finalize_super_block(struct exfat* ef)
{
	if (ef->ro)
		return;

	ef->sb->volume_state = cpu_to_le16(
			le16_to_cpu(ef->sb->volume_state) & ~EXFAT_STATE_MOUNTED);

	/* Some implementations set the percentage of allocated space to 0xff
	   on FS creation and never update it. In this case leave it as is. */
	if (ef->sb->allocated_percent != 0xff)
	{
		uint32_t free, total;

		free = exfat_count_free_clusters(ef);
		total = le32_to_cpu(ef->sb->cluster_count);
		ef->sb->allocated_percent = ((total - free) * 100 + total / 2) / total;
	}

	commit_super_block(ef);	/* ignore return code */
}

void exfat_unmount(struct exfat* ef)
{
	exfat_flush_nodes(ef);	/* ignore return code */
	exfat_flush(ef);		/* ignore return code */
	exfat_put_node(ef, ef->root);
	exfat_reset_cache(ef);
	finalize_super_block(ef);
	exfat_free(ef);			/* will close the descriptor */
}
